tls/ttls/net.c

```diff
@@ -31,547 +32,227 @@
-
-    FD_ZERO( &read_fds );
-    FD_SET( fd, &read_fds );
+	struct socket *sock = ((mbedtls_net_context *)ctx)->socket;
+	struct msghdr msg;
+	struct iovec iov;
+	mm_segment_t oldfs;
+	int ret;
+
+	if(sock == NULL)
+		return MBEDTLS_ERR_NET_INVALID_CONTEXT;
+
+	if(sock->sk == NULL)
+		return MBEDTLS_ERR_NET_INVALID_CONTEXT;
+
+	iov.iov_base = buf;
+	iov.iov_len = len;
+
+	memset(&msg, 0, sizeof(msg));
+	msg.msg_iter.type = ITER_IOVEC;
+	msg.msg_iter.count = len;
+	msg.msg_iter.iov = &iov;
+	msg.msg_iter.nr_segs = 1;
+
+	oldfs = get_fs();
+	set_fs(KERNEL_DS);
+	ret = sock_recvmsg(sock, &msg, len, MSG_DONTWAIT);
+	set_fs(oldfs);
+
+	if(ret < 0) {
+		switch(ret) {
+#if defined EAGAIN
+		case -EAGAIN:
+#endif
+#if defined EWOULDBLOCK && EWOULDBLOCK != EAGAIN
+		case -EWOULDBLOCK:
+#endif
+			return MBEDTLS_ERR_SSL_WANT_READ;
+		}
 
-    tv.tv_sec  = timeout / 1000;
-    tv.tv_usec = ( timeout % 1000 ) * 1000;
+		if(ret == -EPIPE || ret == -ECONNRESET)
+			return MBEDTLS_ERR_NET_CONN_RESET;
 
-    ret = select( fd + 1, &read_fds, NULL, NULL, timeout == 0 ? NULL : &tv );
+		if(ret == -EINTR)
+			return MBEDTLS_ERR_SSL_WANT_READ;
 
-    /* Zero fds ready means we timed out */
-    if( ret == 0 )
-        return( MBEDTLS_ERR_SSL_TIMEOUT );
+		return MBEDTLS_ERR_NET_RECV_FAILED;
+	}
 
-    if( ret < 0 )
-    {
-#if ( defined(_WIN32) || defined(_WIN32_WCE) ) && !defined(EFIX64) && \
-    !defined(EFI32)
-        if( WSAGetLastError() == WSAEINTR )
-            return( MBEDTLS_ERR_SSL_WANT_READ );
-#else
-        if( errno == EINTR )
-            return( MBEDTLS_ERR_SSL_WANT_READ );
-#endif
-
-        return( MBEDTLS_ERR_NET_RECV_FAILED );
-    }
-
-    /* This call will not block */
-    return( mbedtls_net_recv( ctx, buf, len ) );
+	return ret;
 }
+EXPORT_SYMBOL(mbedtls_net_recv);
 
 /*
  * Write at most 'len' characters
  */
-int mbedtls_net_send( void *ctx, const unsigned char *buf, size_t len )
+int
+mbedtls_net_send(void *ctx, const unsigned char *buf, size_t len)
 {
-    int ret;
-    int fd = ((mbedtls_net_context *) ctx)->fd;
-
-    if( fd < 0 )
-        return( MBEDTLS_ERR_NET_INVALID_CONTEXT );
-
-    ret = (int) write( fd, buf, len );
-
-    if( ret < 0 )
-    {
-        if( net_would_block( ctx ) != 0 )
-            return( MBEDTLS_ERR_SSL_WANT_WRITE );
+	struct socket *sock = ((mbedtls_net_context *)ctx)->socket;
+	struct msghdr msg;
+	struct iovec iov;
+	mm_segment_t oldfs;
+	int ret;
+
+	if(sock == NULL)
+		return MBEDTLS_ERR_NET_INVALID_CONTEXT;
+
+	iov.iov_base = (unsigned char *)buf;
+	iov.iov_len = len;
+
+	memset(&msg, 0, sizeof(msg));
+	msg.msg_iter.type = ITER_IOVEC;
+	msg.msg_iter.count = len;
+	msg.msg_iter.iov = &iov;
+	msg.msg_iter.nr_segs = 1;
+
+	oldfs = get_fs();
+	set_fs(KERNEL_DS);
+	ret = sock_sendmsg(sock, &msg);
+	set_fs(oldfs);
+
+	if(ret < 0) {
+		switch(ret) {
+#if defined EAGAIN
+		case -EAGAIN:
+#endif
+#if defined EWOULDBLOCK && EWOULDBLOCK != EAGAIN
+		case -EWOULDBLOCK:
+#endif
```

Comment:
Don't you know what is defined? Usually such ifdefs are used in crossplathform code, not in our case.